### ISSUE TITLE
bug: project's own brigade.js swallows job failures

### ISSUE THREAD

**Issue by krancour** [brigadecore/brigade#921]
The `notificationWrap(...)` helper sandwiches a job between two notifications to Github. If the sandwiched job fails, the second notification communicates that failure to Github, _but as long as that second notification succeeds, the promise that is returned resolves (it's from the second notification; not from the failed job)._

Even if this is corrected, another problem is lurking: the `runSuite(..)` handler is also swallowing errors-- logging them, but not propagating them.

Bottom line is that these two issues currently stand in the way of a build failing when the go or JS unit tests fail.

**Comment by krancour** [comment]
The plot thickens. `runSuite(...)` makes two calls to `checkRun(...)`-- once each for Go tests and JS tests. For both of these calls, `checkRun(...)`, in turn, invokes an `async` function `notificationWrap(...)`.

The long and short of it is that `runSuite(...)` (which is the handler) returns (nothing) while tests are still running. This effectively guarantees that this handler is _always_ considered successful, even while test execution is still in-progress.

I am no JS expert, but I'm thinking that `async` functions probably shouldn't be invoked by a handler.

**Comment by krancour** [comment]
> I am no JS expert, but I'm thinking that async functions probably shouldn't be invoked by a handler.

Ok... after more research, this isn't so much the case, but we _do_ need to make sure the handler will either return a promise or throw an error.

This turns out to be kind of tricky...

Both calls to `checkRun(...)` return a promise, so I was initially thinking something like this could work:

```
function runSuite(e, p) {
  return Promise.all([
    checkRun(e, p, goTest, "go"),
    checkRun(e, p, jsTest, "javascript")
  ]);
```

The problem with this, however, is that as soon as _one_ of the promises from one of the `checkRun(....)` calls is rejected, the promise from `Promise.all(...)` reject. This will properly ensure the whole build fails, however, the worker may terminate before all expected jobs have run (or run to completion). For instance, if the Go tests fail before the job to notify Github of JS test success runs, that notification will never happen. This is clearly not what we want. If something is wrapped in a notification, we want to be sensitive to its failures, but we _also_ want to let all the notification-->job-->notification chains complete.

There seem to be some patterns for making preventing `Promise.all(...)` from failing fast and I am researching those now.

**Comment by krancour** [comment]
Something like this works:

```
function runSuite(e, p) {
  return Promise.all([
    checkRun(e, p, goTest, "go").catch((err) => {return err}),
    checkRun(e, p, jsTest, "javascript").catch((err) => {return err}),
  ])
  .then((values) => {
    values.forEach((value) => {
      if (value instanceof Error) throw value;
    });
  });
}
```

For reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#Promise.all_fail-fast_behaviour
